Feat: Intro screen migration to Jetpack Compose

jackq97 opened the pull request:
Fixes #4254

migrated the intro screen to Jetpack Compose, replacing the previous ViewPager implementation.
- A new `IntroScreen` composable that manages the overall layout and state.
- `IntroPage` composable to display individual intro pages with animated content.
- `PageIndicator` composable for visual page navigation.
- Styling and animations for a smoother user experience.

MohitMaliFtechiz reviewed (changes requested):
@jackq97 Thanks for the PR.

Always mention the issue in every PR, like for this PR Fixes #4254, so that it will link with the relevant ticket, and will close the issue when the PR merge.


The current page indicator animation looks fine, we don't need to change it. There is not much difference in the animation.

jackq97 commented:
> > @MohitMaliDeveloper this is my first draft.
> 
> @jackq97 Please mark your PR as draft until it is ready for review. It will be helpful for us to track that you are currently working on the PR.

yeah sure :)

jackq97 commented:
> Fixes #4254
> 
> migrated the intro screen to Jetpack Compose, replacing the previous ViewPager implementation.
> 
>     * A new `IntroScreen` composable that manages the overall layout and state.
> 
>     * `IntroPage` composable to display individual intro pages with animated content.
> 
>     * `PageIndicator` composable for visual page navigation.
> 
>     * Styling and animations for a smoother user experience.

jackq97 commented:
> @MohitMaliDeveloper this is my first draft. you can give feedback . i would like some info on the page indicator i have added. It's animation is different from the xml one. if you want me to make it exactly like the previous one. we can try that.
> 
> also I'll be working on test cases now.

oh sorry I'll keep it in mind in my next pr. Thanks for confirming the animation. There's one more thing i would like some info on.
I'm writing the test cases and pretty much done with it but I'm unable to figure out how does swiping worked in our xml tests cause I'm unable to remove it and it's causing issues.

```
class IntroRobot : BaseRobot() {
  fun swipeLeft() {
    isVisible(ViewId(R.id.get_started))
    isVisible(TextId(string.welcome_to_the_family))
    isVisible(TextId(string.humankind_knowledge))
    attempt(10) {
      isVisible(TextId(string.save_books_offline))
      isVisible(TextId(string.download_books_message))
    }
    attempt(10) {
      isVisible(TextId(string.save_books_in_desired_storage))
      isVisible(TextId(string.storage_location_hint))
    }
  }

  infix fun clickGetStarted(func: TopLevelDestinationRobot.() -> Unit): TopLevelDestinationRobot {
    testFlakyView({ onView(withId(R.id.get_started)).perform(click()) })
    return topLevel(func)
  }
}
```
so , i can see the `swipeLeft` is a function in Base robot class , but we are not using it anywhere for swiping. and after migrating everything to compose  it still swipes the screen and gives me recyclerview error that's why I'm trying to remove swiping. right now I'm digging classes to find how exactly it's swiping the screen. so if you can point me at the right direction that would be great, meanwhile i'll do my own research .

MohitMaliFtechiz commented:
> so , i can see the swipeLeft is a function in Base robot class , but we are not using it anywhere for swiping. and after migrating everything to compose it still swipes the screen and gives me recyclerview error that's why I'm trying to remove swiping. right now I'm digging classes to find how exactly it's swiping the screen. so if you can point me at the right direction that would be great, meanwhile i'll do my own research .

@jackq97 The swipe automatically happens after 2 seconds, and we are testing after the swipe happens. You can do the same.

jackq97 commented on app/src/main/java/org/kiwix/kiwixmobile/intro/IntroFragment.kt:
```diff
@@ -18,129 +18,56 @@
-            fragmentIntroBinding?.viewPager?.setCurrentItem(currentPage++, true)
-          }
+    composeView.setContent {
+      KiwixTheme {
```
Could you please explain a bit more on what to move to ui screen ??

jackq97 replied:
> @jackq97 I am talking about the `KiwixTheme` and `Surface`. These should be inside the `IntroScreen`.

oh , thanks for clarifying will do.

MohitMaliFtechiz commented:
Hi @jackq97, why have you closed this PR?

Oh, it seems you force-pushed the code, but there are no commits in the PR, so it's closed automatically(I guess this is the same reason for the other PRs that have been closed like this). 

It seems rebase was not properly done.

MohitMaliFtechiz commented:
>  i have noticed we don't have the colors of heading text in compose colors file so right now i have those colors in heading text composable function. we can move it to colors tab if you want.

It would be great to have those in colors.

> with this error the screen glitches and goes to page 1 of intro so I'm unable to validate the text of second page. It is really confusing to me to why no adapter error is happening to a jetpack compose horizontal pager. I have tried checking for other references to some recycler view in test cases but there's none and intro screen is literally the first screen that gets shown. so could you please help out here. i have little code written for debugging the test cases if you want i can push it . cheers

@jackq97 Thanks for debugging and trying different things to fix. I will have a look at this, and will push the fix.

MohitMaliFtechiz commented:
@jackq97, I have made these changes in your existing code.

* Refactored IntroFragmentTest to align with the Compose UI(On which you were facing the issue).
* Fixed: `IntroFragmentTest`, `KiwixSettingsFragmentTest`, and `KiwixSplashActivityTest`.
* Resolved memory leaks in `IntroFragment` and `LanguageFragment` (Reported by the CI, see [in](https://github.com/kiwix/kiwix-android/actions/runs/14580273212/job/40895265399#step:8:1129)).
* A little bit of improvement in `IntroScreen` for better maintaibility which resolved the lint error.

After these changes, is the PR ready for review, or is there a final test and refactoring left to do?

jackq97 commented:
> @jackq97, I have made these changes in your existing code.
> 
>     * Refactored IntroFragmentTest to align with the Compose UI(On which you were facing the issue).
> 
>     * Fixed: `IntroFragmentTest`, `KiwixSettingsFragmentTest`, and `KiwixSplashActivityTest`.
> 
>     * Resolved memory leaks in `IntroFragment` and `LanguageFragment` (Reported by the CI, see [in](https://github.com/kiwix/kiwix-android/actions/runs/14580273212/job/40895265399#step:8:1129)).
> 
>     * A little bit of improvement in `IntroScreen` for better maintaibility which resolved the lint error.
> 
> 
> After these changes, is the PR ready for review, or is there a final test and refactoring left to do?

hi thanks for the changes and fixing the code , I'll check everything and let you know and upgrade my draft for review .

MohitMaliFtechiz commented:
@jackq97 Thanks, please rebase and resolved the conflicts as well.

jackq97 commented:
@MohitMaliDeveloper I've checked the code there's no problem and everything is working fine. I've rebased and solved the merge conflict.
I've checked the test code and i would like more info on how you fixed it? did you figure out why where we getting the  `E No adapter attached; skipping layout` with the compose horizontal pager and how you were able to fix it.
by looking at the code. I think we are manually flipping to the next page and confirming our text nodes on each page.? and by quickly flipping the pages it stops the auto scrolling in our pager so we never encounter that error and we also confirm our nodes.

would love to know more about it.

MohitMaliFtechiz commented:
> @MohitMaliDeveloper I've checked the code there's no problem and everything is working fine. I've rebased and solved the merge conflict.

@jackq97 Thanks for resolving the merge conflicts, but there is still a `Static Analysis` issue. Please fix it.


> I've checked the test code and i would like more info on how you fixed it? did you figure out why were we getting the E No adapter attached; skipping layout with the compose horizontal pager and how you were able to fix it.
by looking at the code. I think we are manually flipping to the next page and confirming our text nodes on each page.? and by quickly flipping the pages it stops the auto scrolling in our pager so we never encounter that error and we also confirm our nodes.

It was not an issue of the `No adapter attached`, it was due to the horizontal pager taking 2 seconds to move to the next page, we were checking the next page content immediately,y that's why the compose was unable to find the second page content. Additionally, adding additional time to wait in test cases using Espresso causes unstable behavior in the test case. That's why I have did this.

jackq97 commented:
> > @MohitMaliDeveloper  
> @jackq97 Thanks for resolving the merge conflicts, but there is still a `Static Analysis` issue. Please fix it.

Hi, the static analysis , 



> > @MohitMaliDeveloper I've checked the code there's no problem and everything is working fine. I've rebased and solved the merge conflict.
> 
> @jackq97 Thanks for resolving the merge conflicts, but there is still a `Static Analysis` issue. Please fix it.
> 
> > I've checked the test code and i would like more info on how you fixed it? did you figure out why were we getting the E No adapter attached; skipping layout with the compose horizontal pager and how you were able to fix it.
> > by looking at the code. I think we are manually flipping to the next page and confirming our text nodes on each page.? and by quickly flipping the pages it stops the auto scrolling in our pager so we never encounter that error and we also confirm our nodes.
> 
> It was not an issue of the `No adapter attached`, it was due to the horizontal pager taking 2 seconds to move to the next page, we were checking the next page content immediately, that's why the compose was unable to find the second page content. Additionally, adding additional time to wait in test cases using Espresso causes unstable behavior in the test case. That's why I have did this.

Hi , i missed an un-used import before pushing the code on weekend. i fixed it.
Thanks  for the explanation it makes sense will keep it in mind.

MohitMaliFtechiz reviewed (approved):
@jackq97 Thanks for the changes. @kelson42 LGTM.
